dhall/src/Dhall/TypeCheck.hs

```diff
@@ -937,86 +941,93 @@ infer typer = loop
+                                Nothing -> do
+                                    let _T'' = quote names _T'
 
-                                return (Just y₀, _T₁')
+                                    die (HandlerNotAFunction y _T'')
 
-            let _T₁'' = quote names _T₁'
+            let matched =
+                    Data.Map.intersectionWithKey match (Dhall.Map.toMap yTs') (Dhall.Map.toMap yUs')
 
-            _ <- loop ctx _T₁''
+            let checkMatched
+                    :: Data.Map.Map Text (Either (TypeError s a) (Val a))
+                    -> Either (TypeError s a) (Maybe (Val a))
+                checkMatched = fmap (fmap snd) . foldlWithKeyM go Nothing
+                  where
+                    go y₁ myT₀' eT₁' = do
+                        _T₁' <- eT₁'
 
-            let process y mU = do
-                    case Dhall.Map.lookup y yTs' of
-                        Nothing -> do
-                            die (MissingHandler diffU)
+                        case myT₀' of
+                            Nothing -> return (Just (y₁, _T₁'))
 
-                        Just _T' -> do
-                            _T₃' <- do
-                                case mU of
-                                    Nothing -> do
-                                        return _T'
+                            Just (y₀, _T₀') -> do
+                                if Eval.conv values _T₀' _T₁'
+                                    then return (Just (y₀, _T₀'))
 
-                                    Just _A₁' -> do
-                                        case Eval.toVHPi _T' of
-                                            Just (x, _A₀', _T₂') -> do
-                                                if Eval.conv values _A₀' _A₁'
-                                                    then do
-                                                        return ()
+                                    else do
+                                        let _T₀'' = quote names _T₀'
+                                        let _T₁'' = quote names _T₁'
+                                        die (HandlerOutputTypeMismatch y₀ _T₀'' y₁ _T₁'')
 
-                                                    else do
-                                                        let _A₀'' = quote names _A₀'
-                                                        let _A₁'' = quote names _A₁'
+            mT₀' <- checkMatched matched
 
-                                                        die (HandlerInputTypeMismatch y _A₁'' _A₀'')
+            case mT₀' of
+                Nothing -> return ()
 
-                                                return (_T₂' (fresh ctx x))
+                Just _T₀' -> do
+                    let _T₀'' = quote names _T₀'
 
-                                            Nothing -> do
-                                                let _T'' = quote names _T'
+                    _ <- loop ctx _T₀''
```

Comment:
The free variable check here could alternatively be included in `match` above. That might slightly improve type errors like this one:

```
$ echo 'merge { a = None Bool, x = None } (<a | x: Type>.x Bool)' | dhall:dhall type
Error: Handlers should have the same output type

…
- Bool
+ A@-1

1│ merge { a = None Bool, x = None } (<a | x: Type>.x Bool)
```